README

dsniff-2.4b1
------------

This is a fork of Dug Song's dsniff tools, updated to compile and run on OS X.

Installation on OS X:

1. Install dependencies:

brew install berkeley-db libnet libnids libpcap openssl

2. Configure with correct paths:

./configure \
  --with-libpcap=/usr/local/opt/libpcap \
  --with-openssl=/usr/local/opt/openssl \
  --with-libnet=/usr/local/opt/libnet \
  --with-libnids=/usr/local/opt/libnids \
  --without-x

3. Compile:

make

4. Install:

make install

------------

Original README follows

------------

i wrote these tools with honest intentions - to audit my own network,
and to demonstrate the insecurity of cleartext / weakly-encrypted
network protocols and ad-hoc PKI. please do not abuse this software.

these programs require:

      Berkeley DB - http://www.sleepycat.com/
      OpenSSL - http://www.openssl.org/
      libpcap - http://www.tcpdump.org/
      libnids - http://www.packetfactory.net/Projects/Libnids/
      libnet - http://www.packetfactory.net/Projects/Libnet/

built and tested on OpenBSD, Linux, and Solaris. YMMV.

what's here:

arpspoof
	redirect packets from a target host (or all hosts) on the LAN
	intended for another local host by forging ARP replies. this
	is an extremely effective way of sniffing traffic on a switch.
	kernel IP forwarding (or a userland program which accomplishes
	the same, e.g. fragrouter :-) must be turned on ahead of time.

dnsspoof
	forge replies to arbitrary DNS address / pointer queries on
	the LAN. this is useful in bypassing hostname-based access
	controls, or in implementing a variety of man-in-the-middle
	attacks (HTTP, HTTPS, SSH, Kerberos, etc).

dsniff
	password sniffer. handles FTP, Telnet, SMTP, HTTP, POP,
	poppass, NNTP, IMAP, SNMP, LDAP, Rlogin, RIP, OSPF, PPTP
	MS-CHAP, NFS, VRRP, YP/NIS, SOCKS, X11, CVS, IRC, AIM, ICQ,
	Napster, PostgreSQL, Meeting Maker, Citrix ICA, Symantec
	pcAnywhere, NAI Sniffer, Microsoft SMB, Oracle SQL*Net, Sybase
	and Microsoft SQL auth info.

	dsniff automatically detects and minimally parses each
	application protocol, only saving the interesting bits, and
	uses Berkeley DB as its output file format, only logging
	unique authentication attempts. full TCP/IP reassembly is
	provided by libnids(3) (likewise for the following tools as
	well).

filesnarf
	saves selected files sniffed from NFS traffic in the current
	working directory.

macof
	flood the local network with random MAC addresses (causing
	some switches to fail open in repeating mode, facilitating
	sniffing). a straight C port of the original Perl Net::RawIP
	macof program.

mailsnarf
	a fast and easy way to violate the Electronic Communications
	Privacy Act of 1986 (18 USC 2701-2711), be careful. outputs
	selected messages sniffed from SMTP and POP traffic in Berkeley
	mbox format, suitable for offline browsing with your favorite
	mail reader (mail -f, pine, etc.).

msgsnarf
	record selected messages from sniffed AOL Instant Messenger,
	ICQ 2000, IRC, and Yahoo! Messenger chat sessions.

sshmitm
	SSH monkey-in-the-middle. proxies and sniffs SSH traffic
	redirected by dnsspoof(8), capturing SSH password logins, and
	optionally hijacking interactive sessions. only SSH protocol
	version 1 is (or ever will be) supported - this program is far
	too evil already.

sshow
	SSH traffic analysis tool. analyzes encrypted SSH-1 and SSH-2
	traffic, identifying authentication attempts, the lengths of
	passwords entered in interactive sessions, and command line
	lengths.

tcpkill
	kills specified in-progress TCP connections (useful for
	libnids-based applications which require a full TCP 3-whs for
	TCB creation).

tcpnice
	slow down specified TCP connections via "active" traffic
	shaping. forges tiny TCP window advertisements, and optionally
	ICMP source quench replies.

urlsnarf
	output selected URLs sniffed from HTTP traffic in CLF
	(Common Log Format, used by almost all web servers), suitable
	for offline post-processing with your favorite web log
	analysis tool (analog, wwwstat, etc.).

webmitm
	HTTP / HTTPS monkey-in-the-middle. transparently proxies and
	sniffs web traffic redirected by dnsspoof(8), capturing most
	"secure" SSL-encrypted webmail logins and form submissions.

webspy
	sends URLs sniffed from a client to your local Netscape
	browser for display, updated in real-time (as the target
	surfs, your browser surfs along with them, automagically).
	a fun party trick. :-)

-d.

---
http://www.monkey.org/~dugsong/


CHANGES
$Id: CHANGES,v 1.56 2001/03/19 06:53:47 dugsong Exp $

- Add support for ICMP frag-needed to tcpnice.

- New sshow program:
  http://www.openwall.com/advisories/OW-003-ssh-traffic-analysis.txt

- Add support for HTTP absolute URIs to webmitm and urlsnarf.

- Add support for non-transparent proxying to webmitm, as requested
  by Juergen Schmidt <ju@ct.heise.de>.

- Fix TDS decode for MSSQL 7.x, from Paul van Maaren
  <P.v.Maaren@reseau.nl>.

- Kludge around stupid Redhat Linux 6, 7 build env inconsistencies.

- Fix dumb endianness bug in dnsspoof.


v2.3 Sun Dec 17 11:35:38 EST 2000

- Add VRRP parsing to dsniff, from Eric Jackson <shinobi@monkey.org>.

- Require pcap filter argument for tcpkill, tcpnice.

- Add Microsoft PPTP MS-CHAP (v1, v2) parsing to dsniff, based on
  anger.c by Aleph One <aleph1@securityfocus.com>.

- Fix pcAnywhere 7, 9.x parsing in dsniff.

- Add -t trigger[,...] flag to dsniff, to specify individual triggers
  on the command line.

- Convert most everything to use new buf interface.

- New programs: dnsspoof, msgsnarf, sshmitm, webmitm.

- Fix inverted regex matching in *snarf programs.

- Consistent arpspoof, macof, tcpnice, tcpkill output.

- Rename arpredirect to arpspoof (maintain consistent *sniff, *snarf,
  *spoof, *spy nomenclature).

- Consistent pcap filter argument to dsniff, *snarf programs.

- Add trigger for Checkpoint Firewall-1 Session Authentication Agent
  (261/tcp), as suggested by Joe Segreti <seg@clark.net>.

- Add SMTP parsing to dsniff, as requested by Denis Ducamp
  <Denis.Ducamp@hsc.fr>.

- Add rexec and RPC ypserv parsing to dsniff, as requested by 
  Oliver Friedrichs <of@securityfocus.com>.

- Add HTTP proxy auth parsing back to dsniff, it got lost in the
  shuffle. Reported by Denis Ducamp <Denis.Ducamp@hsc.fr>.

- Add NNTPv2 and other AUTHINFO extensions to dsniff.


v2.2 Wed Jun 14 00:58:37 EDT 2000

- Rewrite HTTP decoding in dsniff, adding support for QUERY_STRING and
  x-www-form-urlencoded parsing (various CGI authentication schemes).
  
- Alpha support (libnids and libnet still need to be fixed).

- Fix arp discovery in arpredirect on Linux.

- Add -m flag to enable automatic protocol detection in dsniff,
  based on the classic file(1) command by Ian Darwin.

- Add TDS (Sybase, Microsoft SQL Server) parsing to dsniff.

- Clean up RPC decodes, TCP half-duplex reassembly in dsniff.

- New filesnarf program.

- Add regular expression matching to mailsnarf.

- Add POP support to mailsnarf.


v2.1 Thu May 18 16:18:35 EDT 2000

- Add -c flag to specify half-duplex TCP stream reassembly in dsniff
  (better support for sniffing off switched ports using arpredirect).

- Fix > 24 char Meeting Maker passwd parsing in dsniff.

- Fix OSPF parsing in dsniff (don't truncate first two chars),
  as reported by Felix Contreras <cfelix@fisiologia.com>.

- Fix webspy URL ignoring, as reported by Interrupt <mike@eEye.com>.


v2.0 Tue May 16 13:11:22 EDT 2000

- Major dsniff rewrite, since ppl are actually reading this code. :-)

- Add configurable decode triggers to dsniff.

- Add dsniff debugging functions, split out decode routines.

- Add yppasswd parsing to dsniff.

- Rewrite dsniff RPC framework, portmap and NFS mountd decodes.

- Make dsniff savefile format portable.

- Remove findgw - to be subsumed by dsquat package.

- Add PostgreSQL parsing to dsniff.

- Add Meeting Maker parsing to dsniff.

- Add poppass parsing to dsniff.

- Add RIP, OSPF parsing to dsniff.

- Fix RSET handling in mailsnarf (from Martin Fredriksson <martin@crt.se>).


v1.8 Sun Apr  9 23:59:46 EDT 2000

- Add SOCKS parsing to dsniff.

- Add pcAnywhere parsing to dsniff.

- Fix SMB parsing in dsniff.

- Add IRC parsing to dsniff.

- Add NAI Sniffer parsing to dsniff (from Anonymous).


v1.7 Mon Mar 27 16:19:32 EST 2000

- Add -s flag to specify snaplen to dsniff.

- Support systems without <libgen.h> or dirname().

- Add Microsoft SMB parsing to dsniff.

- Add Citrix ICA parsing to dsniff.

- Add LDAP parsing to dsniff.

- Fix Berkeley mbox format again (\n, not \r\n).

- Fix null URI dereference in urlsnarf.

- Add Oracle SQL*Net (v2, Net8) parsing to dsniff.

- Catch data left on connection close in mailsnarf, urlsnarf, webspy.


v1.6 Sun Mar 12 16:25:09 EST 2000

- Support non-glibc Linux systems missing ether_ntoa().

- Unique HTTP auth info by URI dirname in dsniff.

- Add Napster parsing to dsniff.

- Don't rely on /etc/services for dsniff.

- Add AIM, ICQ (v2, v5) parsing to dsniff.

- Add CVS pserver parsing to dsniff.

- Skip IMAP command tag in dsniff.


v1.5 Tue Feb 15 23:22:25 EST 2000

- Fix HTTP proxy support in urlsnarf (from <felix@convergence.de>).

- Fix HTTP proxy support in dsniff (from <Alain.Thivillon@hsc.fr>).

- Proper manpages for all programs.

- Strip binary nulls in telnet input, in dsniff (doh!).


v1.4 Thu Jan 27 12:08:41 EST 2000

- Add verbose flag (-v) to tcpkill, tcpnice.

- Add NNTP parsing to dsniff (from Felix von Leitner <felix@convergence.de>).

- Fix mailsniff mbox formatting of ^From in message body.

- Add HTTP proxy support in dsniff, urlsnarf, webspy.

- Fix getopt() usage to be POSIX compliant (s/EOF/-1/).

- New tcpnice program.


v1.3 Fri Jan 21 02:47:37 EST 2000

- Ported to Solaris (along with libnids :-)

- Add Berkeley db(3) output file format to dsniff, as well as
  restricting logging to unique auth info.

- New tcpkill program.

- New lame dsniff(8) manpage.

- Add DNS lookups (and -n flag to disable) in dsniff, urlsnarf.

- Add HTTP Basic Authentication, Referer, User-Agent logging to urlsnarf.

- Improve RPC message parsing in dsniff.

- Improve SMTP parsing in mailsnarf.

- Improve HTTP 1.x parsing in dsniff, urlsnarf, webspy.

- Fix IMAP, Rlogin, Telnet option parsing in dsniff (broke them in 1.2).

- Add X11 MIT-MAGIC-COOKIE parsing to dsniff.

- Don't forget to decode POP SASL username in dsniff (doh!).


v1.2 Sat Jan  8 22:36:42 EST 2000

- Ported to FreeBSD (but not tested!).

- Add GNU autoconf support.


ssh.h
/*
 * ssh.h
 *
 * Minimal SSH-1 protocol implementation.
 *
 * Copyright (c) 2000 Dug Song <dugsong@monkey.org>
 *
 * $Id: ssh.h,v 1.3 2001/03/15 08:33:06 dugsong Exp $
 */

#ifndef SSH_H
#define SSH_H

#define SSH_MAX_PADLEN		8
#define SSH_MAX_PKTLEN		262144

/* Authentication types. */
#define SSH_AUTH_PASSWORD	3

/* Cipher types. */
#define SSH_CIPHER_NONE		0
#define SSH_CIPHER_3DES		3
#define SSH_CIPHER_BLOWFISH	6

/* Message types. */
#define SSH_MSG_DISCONNECT	1
#define SSH_SMSG_PUBLIC_KEY	2
#define SSH_CMSG_SESSION_KEY	3
#define SSH_CMSG_USER		4
#define SSH_CMSG_AUTH_PASSWORD	9
#define SSH_SMSG_SUCCESS	14
#define SSH_SMSG_FAILURE	15
#define SSH_CMSG_STDIN_DATA	16
#define SSH_SMSG_STDOUT_DATA	17
#define SSH_SMSG_STDERR_DATA	18
#define SSH_SMSG_EXITSTATUS	20

typedef struct ssh_ctx {
	int	 authmask;
	int	 encmask;
	RSA	*servkey;
	RSA	*hostkey;
} SSH_CTX;

typedef struct ssh {
	int	 fd;
	SSH_CTX *ctx;
	u_char	 sesskey[32];
	void	*estate;
	void	*dstate;
	void   (*encrypt)(u_char *src, u_char *dst, int len, void *state);
	void   (*decrypt)(u_char *src, u_char *dst, int len, void *state);
} SSH;

void	 SSH_init(void);

SSH_CTX	*SSH_CTX_new(void);

SSH	*SSH_new(SSH_CTX *ctx);

void	 SSH_set_fd(SSH *ssh, int fd);

int	 SSH_accept(SSH *ssh);

int	 SSH_connect(SSH *ssh);

int	 SSH_recv(SSH *ssh, u_char *buf, int size);

int	 SSH_send(SSH *ssh, u_char *buf, int len);

void	 SSH_close(SSH *ssh);

#endif /* SSH_H */
